### PR TITLE
[Merged by Bors] - feat: port Tactic.Slice

### LINKED ISSUE #2099: port `slice` tactics
We need to port `slice`, `slice_lhs`, and `slice_rhs`.

### PR DESCRIPTION
**Pull request by mattrobball**
Ports the `slice` and its descendants. The new syntax is `slice_lhs a b => tac`. 

Beyond a new `Slice.lean` and a test file, there are changes to `Mathlib.Tactic.Core` adding iteration tactics and replacing a `MonadExceptOf` instance assumption by a `MonadExcept` instance for these tactics.

### CODE REVIEW

File: Mathlib/Tactic/Core.lean [hunk 1]
```diff
@@ -125,18 +132,53 @@ def allGoals (tac : TacticM Unit) : TacticM Unit := do
 def andThenOnSubgoals (tac1 : TacticM Unit)  (tac2 : TacticM Unit) : TacticM Unit :=
   focus do tac1; allGoals tac2
 
-variable [Monad m] [MonadExceptOf Exception m]
+variable [Monad m] [MonadExcept Exception m]
```
Comment (mattrobball) [inline Mathlib/Tactic/Core.lean:135]: This change did not break anything but it would benefit from input from an expert.

**Comment by mattrobball** [comment]
#1287 now depends on this PR. It also serves a useful test for the correctness of the implementation.

File: Mathlib/Tactic/Slice.lean [hunk 2]
```diff
@@ -49,25 +55,25 @@ def evalSlice (a b : Nat) : TacticM Unit := do
...
-/-- 
-`sliceLHS a b => tac` is a conv tactic which enters the LHS of a target, 
-uses `slice` to extract the `a` through `b` terms, and then applies 
-`tac` to the result. 
+/--
+`slice_lhs a b => tac` zooms to the left hand side, uses associativity for categorical
+composition as needed, zooms in on the `a`-th through `b`-th morphisms, and invokes `tac`.
 -/
-syntax (name := sliceLHS) "sliceLHS" num num " => " convSeq : tactic
+syntax (name := sliceLHS) "slice_lhs " num num " => " convSeq : tactic
```
Comment (mattrobball) [inline Mathlib/Tactic/Slice.lean:65]: This conflicts with the naming convention for tactics in Lean core. Eg. [Lean.Init.Tactics](https://leanprover-community.github.io/mathlib4_docs/Init/Tactics.html). Though mathlib seems to have not adhered to any one convention...

Comment (thorimur) [reply Mathlib/Tactic/Slice.lean:65]: For porting tactics, we use the syntax given in [`Mathlib/Mathport/Syntax.lean`](https://github.com/leanprover-community/mathlib4/blob/ecd37441047e490ff2ad339e16f45bb8b58591bd/Mathlib/Mathport/Syntax.lean#L450), which is what `Mathport` uses!

Also I think the convention for the actual call syntax is lower_snake_case, judging by the syntax for many tactics in that file? I'm not sure if it's written down somewhere though.

Comment (mattrobball) [reply Mathlib/Tactic/Slice.lean:65]: Ah you are right. I am confusing the name with the syntax. Thanks!

**Comment by rwbarton** [comment]
Thanks!
bors merge

**Pull request closed without merging**